test/e2e/network/dual_stack.go

```diff
@@ -765,6 +717,33 @@ func validateServiceAndClusterIPFamily(svc *v1.Service, expectedIPFamilies []v1.
 	}
 }
 
+func validateEndpointsAndEndpointSlices(ctx context.Context, f *framework.Framework, svc *v1.Service, expectedIPFamilies []v1.IPFamily) {
+	var endpoint *v1.Endpoints
+	if err := wait.PollUntilContextTimeout(ctx, 500*time.Millisecond, 10*time.Second, true, func(ctx context.Context) (bool, error) {
+		var err error
+		endpoint, err = f.ClientSet.CoreV1().Endpoints(svc.Namespace).Get(ctx, svc.Name, metav1.GetOptions{})
+		return err == nil, nil
+	}); err != nil {
+		framework.Failf("Get endpoints for service %s/%s failed (%s)", svc.Namespace, svc.Name, err)
+	}
+	framework.Logf("Got endpoint %#v", endpoint)
+	validateEndpointsBelongToIPFamily(svc, endpoint, expectedIPFamilies[0] /* endpoint controller works on primary IP */)
+
+	var slices *discoveryv1.EndpointSliceList
+	if err := wait.PollUntilContextTimeout(ctx, 500*time.Millisecond, 10*time.Second, true, func(ctx context.Context) (bool, error) {
+		var err error
+		slices, err = f.ClientSet.DiscoveryV1().EndpointSlices(svc.Namespace).List(ctx, metav1.ListOptions{LabelSelector: discoveryv1.LabelServiceName + "=" + svc.Name})
+		if err != nil || len(slices.Items) == 0 {
+			return false, nil
+		}
+		return true, nil
```

Comment:
This may return one EndpointSlice while another EndpointSlice is still being created as EndpointSlice controller creates them sequentially, which may explain the failure of https://github.com/kubernetes/kubernetes/pull/128898#issuecomment-2489937596?

Reply:
yeah, I remember when Rob Scott started adding the endpoitnslices that slices has some time of convergence and there are some cases like that

Reply:
Oh... I was thinking that since the controller always creates the IPv4 and IPv6 slices at the same time, that we didn't have to worry about that, but you're right, just because they're created at the same time doesn't mean they'll get processed by the Watch / Informer at the same time.